vnext/Desktop/Modules/WebSocketModule.cpp

```diff
@@ -13,123 +14,177 @@
+        }
+      }),
+    Method(
+      "close",
+      [this](dynamic args) // [int64_t code, string reason,] int64_t id
+      {
+        // See react-native\Libraries\WebSocket\WebSocket.js:_close
+        if (args.size() == 3) // WebSocketModule.close(statusCode, closeReason, this._socketId);
+        {
+          auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 2));
+          if (auto sharedWs = weakWs.lock())
+          {
+            sharedWs->Close(static_cast<IWebSocketResource::CloseCode>(jsArgAsInt(args, 0)), jsArgAsString(args, 1));
+          }
+        }
+        else if (args.size() == 1) // WebSocketModule.close(this._socketId);
+        {
+          auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 0));
+          if (auto sharedWs = weakWs.lock())
+          {
+            sharedWs->Close(IWebSocketResource::CloseCode::Normal, {});
+          }
+        }
+        else
+        {
+          auto errorObj = dynamic::object("id", -1)("message", "Incorrect number of parameters");
+          this->SendEvent("websocketFailed", std::move(errorObj));
+        }
+      }),
+    Method(
+      "send",
+      [this](dynamic args) // const string& message, int64_t id
+      {
+        auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 1));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Send(jsArgAsString(args, 0));
+        }
+      }),
+    Method(
+      "sendBinary",
+      [this](dynamic args) // const string& base64String, int64_t id
+      {
+        auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 1));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->SendBinary(jsArgAsString(args, 0));
+        }
+      }),
+    Method(
+      "ping",
+      [this](dynamic args) // int64_t id
+      {
+        auto weakWs = this->GetOrCreateWebSocket(jsArgAsInt(args, 0));
+        if (auto sharedWs = weakWs.lock())
+        {
+          sharedWs->Ping();
+        }
+      })
+  };
 } // getMethods
 
 #pragma region private members
 
-void WebSocketModule::SendEvent(string &&eventName, dynamic &&args) {
+void WebSocketModule::SendEvent(string&& eventName, dynamic&& args)
+{
   auto instance = this->getInstance().lock();
   if (instance)
+  {
     instance->callJSFunction("RCTDeviceEventEmitter", "emit", dynamic::array(std::move(eventName), std::move(args)));
+  }
 }
 
-IWebSocketResource *WebSocketModule::GetOrCreateWebSocket(int64_t id, string &&url) {
-  IWebSocketResource *ptr = nullptr;
-  if (m_webSockets.find(id) == m_webSockets.end()) {
+std::weak_ptr<IWebSocketResource> WebSocketModule::GetOrCreateWebSocket(int64_t id, string&& url)
+{
+  if (m_webSockets.find(id) == m_webSockets.end())
+  {
     auto ws = IWebSocketResource::Make(std::move(url));
-    ws->SetOnError([this, id](const IWebSocketResource::Error &err) {
+    ws->SetOnError([this, id, ws](const IWebSocketResource::Error& err)
+    {
+      auto wsRef = ws;
```

Comment:
>auto wsRef = ws; [](start = 6, length = 16)

What does this do? The shared_ptr should already be kept alive until the lambda completes because it captures by copy.

Reply:
True. That's redundant.
Reverting...